File: javascript/스킬트리.js
function verify(x) {
  let cnt = x.filter((z) => z >= 0).length;
  if (cnt === 0) return true;
  let pre = x[0],
    i;
  for (i = 1; i < x.length; i++) {
    if (pre === -1) return false;
    cnt--;
    if (cnt === 0) return true;
    if (x[i] === -1) return false;
    if (pre > x[i]) return false;
    pre = x[i];
  }
  return true;
}
function solution(skill, skill_trees) {
  var answer = 0;
  const s = skill.split('');
  for (let i in skill_trees) {
    let a = [];
    for (let j in s) {
      a.push(skill_trees[i].indexOf(s[j]));
    }
    if (verify(a)) answer++;
  }
  return answer;
}

File: java/스킬트리.java
import java.util.*;
class Solution {
    HashMap<Character,Integer> map  = new HashMap<>();
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        for(int i=0;i<skill.length();i++){
            map.put(skill.charAt(i),i+1);
        }
        for(int i=0;i<skill_trees.length;i++){
            ArrayList<Integer> arr = new ArrayList<>();
            String cur = skill_trees[i];
            for(int j=0;j<cur.length();j++){
                if(map.containsKey(cur.charAt(j))){
                    arr.add(map.get(cur.charAt(j)));
                }
            }
            
            boolean flag=true;
            if(arr.size()>0){
                if(arr.get(0)!=1){
                    flag=false;
                }
            }
            for(int j=1;j<arr.size();j++){
                if(arr.get(j)-arr.get(j-1)!=1){
                    flag=false;
                    break;
                }
            }
            if(flag)
                answer++;
        }
        return answer;
    }
}